src/Microsoft.AspNetCore.Routing/EndpointConstraints/EndpointSelector.cs

```diff
@@ -0,0 +1,442 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing.Matchers;
+using Microsoft.Extensions.Internal;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+
+namespace Microsoft.AspNetCore.Routing.EndpointConstraints
+{
+    internal class EndpointSelector
+    {
+        private static readonly IReadOnlyList<Endpoint> EmptyEndpoints = Array.Empty<Endpoint>();
+
+        private readonly CompositeEndpointDataSource _dataSource;
+        private readonly EndpointConstraintCache _endpointConstraintCache;
+        private readonly ILogger _logger;
+
+        private Cache _cache;
+
+        public EndpointSelector(
+            CompositeEndpointDataSource dataSource,
+            EndpointConstraintCache endpointConstraintCache,
+            ILoggerFactory loggerFactory)
+        {
+            _dataSource = dataSource;
+            _logger = loggerFactory.CreateLogger<EndpointSelector>();
+            _endpointConstraintCache = endpointConstraintCache;
+        }
+
+        private Cache Current
+        {
+            get
+            {
+                var cache = Volatile.Read(ref _cache);
+
+                if (cache != null)
+                {
+                    return cache;
+                }
+
+                cache = new Cache(_dataSource.Endpoints);
+                Volatile.Write(ref _cache, cache);
+                return cache;
+            }
+        }
+
+        public IReadOnlyList<Endpoint> SelectCandidates(RouteContext context)
```

Comment:
`SelectCandidates` is not used at the moment. Does it have a place in dispatching?

Reply:
No, we won't need this. This is part of action selection for convention routing. Dispatcher works like attribute routing